### PR TITLE
feat(github-actions): support GitHub actions runners

### LINKED ISSUE #19827: Update GitHub actions runner automatically
### What would you like Renovate to be able to do?

GitHub has been bumping [Ubuntu](https://github.blog/changelog/2022-08-09-github-actions-the-ubuntu-18-04-actions-runner-image-is-being-deprecated-and-will-be-removed-by-12-1-22/) and [Mac](https://github.blog/changelog/2022-07-20-github-actions-the-macos-10-15-actions-runner-image-is-being-deprecated-and-will-be-removed-by-8-30-22/) runners recently. It would be nice if in these cases we get Renovate PRs, manually looking for these in every single repo is time consuming and error prone.

I can't think of a better way of explaining it than with a visual diff:
```diff
jobs:
  job-name:
    name: My Job
-    runs-on: ubuntu-20.04
+    runs-on: ubuntu-22.04
```

### If you have any ideas on how this should be implemented, please tell us here.

The [documentation](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) is one source data. and [this section](https://github.com/actions/runner-images#available-images) is the source of truth, I think, sadly, this is still just a `README.md`.

They communicate changes in [Announcements](https://github.com/actions/runner-images/issues?q=is%3Aissue+is%3Aopen+label%3AAnnouncement), and [blog posts](https://github.blog/changelog/label/actions/).

This doesn't bode well for a data source, as I personally couldn't find a public resource where the versions are cleanly available, but I noticed you have [static data](https://github.com/renovatebot/renovate/blob/main/data/ubuntu-distro-info.json) (for these cases?) which is pretty close to what would be necessary. I'm sure that the community will be happy to provide updates whenever runners move ahead a version (i.e. e.g. `ubuntu-latest` changes meaning), which should be once every 1-4 years as far as I can see.

### Is this a feature you are interested in implementing yourself?

No

### PR DESCRIPTION
**Pull request by PeterNitsche**
## Changes

- Enhanced GitHub Actions manager to extract used GitHub runners
- Added datasource with available GitHub runners
- Added unit tests for manager and datasource

## Context

Closes https://github.com/renovatebot/renovate/issues/19827

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/github-actions/__fixtures__/workflow_5.yml
Comment (viceice) [inline lib/modules/manager/github-actions/__fixtures__/workflow_5.yml:1]: please inline this small fixture

File: lib/modules/manager/github-actions/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.spec.ts:290]: no new snapshot tests, use toEqual or toMatchObject.

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.ts:91]:
```suggestion
  const runners: string[] = [];
```

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.ts:99]: use regex and filter expressions, they can be matrix job references.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/github-runners/index.ts
Comment (viceice) [inline lib/modules/datasource/github-runners/index.ts:14]:
```suggestion
  private static readonly releases: { [packageName: string]: Release[] } = {
```

**Comment by HonkingGoose** [comment]
I'll just say that we need to document the behavior in the docs.  Even more so if we manually update the list. Or we'll get questions like "why is Renovate not updating me to new version `x`?".

File: lib/modules/manager/github-actions/__fixtures__/workflow_5.yml
Comment (TWiStErRob) [inline lib/modules/manager/github-actions/__fixtures__/workflow_5.yml:7]: Is this invalid config? Duplicate `runs-on`.

Comment (TWiStErRob) [inline lib/modules/manager/github-actions/__fixtures__/workflow_5.yml:7]: There are more valid ways to define this:

```
runs-on: 'ubuntu-22.04'
runs-on: >
  ubuntu-22.04
runs-on: |
  ubuntu-22.04
```

Is the parser automatically taking care of all the ways?

**Review by TWiStErRob — COMMENTED** [review]
Thank you for picking it up!

**Comment by PeterNitsche** [comment]
> I'll just say that we need to document the behavior in the docs. Even more so if we manually update the list. Or we'll get questions like "why is Renovate not updating me to new version `x`?".

@HonkingGoose Absolutely! I was struggling to find the right file to add this kind of documentation. Could you suggest one?

**Comment by PeterNitsche** [comment]
Apologies! I have pushed my changes too early by habit. I am going to rerequest review once I am fully done.

**Comment by viceice** [comment]
> > I'll just say that we need to document the behavior in the docs. Even more so if we manually update the list. Or we'll get questions like "why is Renovate not updating me to new version `x`?".
> 
> @HonkingGoose Absolutely! I was struggling to find the right file to add this kind of documentation. Could you suggest one?

it should be in the github-runners datasource readme

File: lib/modules/manager/github-actions/extract.ts
Comment (PeterNitsche) [reply lib/modules/manager/github-actions/extract.ts:99]: I don't fully understand what you want to achieve, but I will try..

Comment (TWiStErRob) [reply lib/modules/manager/github-actions/extract.ts]: @PeterNitsche `runs-on: ${{ matrix.foo-bar }}` is also possible, and that would parse wrongly with this simple split.

https://github.com/TWiStErRob/detekt-4165/blob/main/.github/workflows/CI.yml#L8
I guess this is not expected to be supported, just an example the complexity.

File: lib/modules/datasource/github-runners/readme.md
Comment (HonkingGoose) [inline lib/modules/datasource/github-runners/readme.md:7]:
```suggestion
New _stable_ runner versions must be added to the datasource with a pull request.
```

File: lib/modules/manager/github-actions/extract.ts
Comment (HonkingGoose) [inline lib/modules/manager/github-actions/extract.ts:91]: Please refactor this `const` to `runnerVersionRegex`.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/modules/manager/github-actions/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/github-actions/readme.md:29]:
```suggestion
For example, Renovate ignores the runner configured in the `RUNNER` variable:
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Two more changes. 😄

**Comment by PeterNitsche** [comment]
I have no idea why the codecov checks failed. Please help.

File: lib/modules/datasource/github-runners/readme.md
Comment (HonkingGoose) [inline lib/modules/datasource/github-runners/readme.md:10]:
```suggestion
Unstable runners are tagged as `[beta]` in the readme of the [`runner-images` repository](https://github.com/actions/runner-images).
Once a runner version becomes stable, the `[beta]` tag is removed and the suffix `latest` is added to its YAML label.
```

We follow the one sentence per line pattern in Markdown files.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/github-runners/readme.md
Comment (PeterNitsche) [reply lib/modules/datasource/github-runners/readme.md:10]: Is there a style guide for those rules?

Comment (HonkingGoose) [reply lib/modules/datasource/github-runners/readme.md]: Good question! You're a new contributor, so I guess some of my style changes feel arbitrary.

I created a style guide a while back, but forgot to add the "Use one sentence per line" for Markdown files rule to it. I'll make a PR to update the guide now. 😉 

Here's our [style guide](https://github.com/renovatebot/renovate/blob/main/docs/development/style-guide.md).

File: lib/modules/datasource/github-runners/index.ts
Comment (viceice) [inline lib/modules/datasource/github-runners/index.ts:33]:
```suggestion
      releases: GithubRunnersDatasource.releases[packageName] ?? [],
```
the else path is probably not taken in tests

https://app.codecov.io/gh/renovatebot/renovate/pull/23633/blob/lib/modules/datasource/github-runners/index.ts#L33

this API should return null instead of empty results.

File: lib/modules/manager/github-actions/readme.md
Comment (viceice) [inline lib/modules/manager/github-actions/readme.md:40]: I see no test for this case

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/github-actions/extract.ts
Comment (HonkingGoose) [reply lib/modules/manager/github-actions/extract.ts:14]: I thought it would be good to use good straight-forward names, so you don't have to guess what a variable means.

Cleaning up as you go seems okay to me. But maybe the maintainers want to rename the variable in the source file?

I'll defer to the Renovate maintainers on this matter.

I don't mean for my review comments to annoy you or to slow down the merge process.

Comment (PeterNitsche) [reply lib/modules/manager/github-actions/extract.ts:14]: I agree, `Regex` is more readable than `Re`. Your review comments were helpful to improve the quality of the PR. Also, your explanations are detailed enough to understand the intention of the change request.  👍 Thanks, @HonkingGoose!

**Comment by PeterNitsche** [comment]
@viceice Could you please approve? There is one remaining discussion and I hope you can agree to keep the change as it is. Thanks!

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [reply lib/modules/manager/github-actions/extract.ts:14]: We don't like to have unrelated changes on feature or bugfix PR's. Those changes should always be on separate PR's. Especially if they are so many, they make review a lot harder than necessary.

Comment (PeterNitsche) [reply lib/modules/manager/github-actions/extract.ts]: Alright, I have renamed the variables to its original names since I don't see another possibility to get this PR merged. 
I don't want to overwhelm the maintainers with 8 additional lines of changed code.

In the future, I will consider other open source projects which align more with my software development ideas.

File: lib/modules/manager/github-actions/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.spec.ts:104]: I think we should add a skip reason, because we can't update latest values. it would print a datasource warning on lockup.

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.ts:92]:
```suggestion
  /^\s*(?<depName>[a-z]+)-(?<currentValue>[^\s]+)/
```
I think we can make it more strict. runner name can only be a-z.

maybe we need to support different cases too, as some users prefer MacOS. not sure if it's supported by GitHub.

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.ts:104]: please extract the callback to a separate function

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.ts:119]: can new value really be null or undefined?

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.ts:115]: use string template instead

File: lib/modules/manager/github-actions/extract.ts
Comment (TWiStErRob) [reply lib/modules/manager/github-actions/extract.ts]: What will this do on `macos-13-xl`?
https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#choosing-github-hosted-runners

File: lib/modules/manager/github-actions/extract.spec.ts
Comment (TWiStErRob) [inline lib/modules/manager/github-actions/extract.spec.ts:4]: What's the behavior when someone uses this syntax?
(Renovate can't update it, but should fail horribly either 🤓)
https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-combining-groups-and-labels

Comment (PeterNitsche) [reply lib/modules/manager/github-actions/extract.spec.ts:4]: Nice catch! I have added a test case to confirm that such use cases are ignored.

File: lib/modules/datasource/github-runners/index.ts
Comment (viceice) [inline lib/modules/datasource/github-runners/index.ts:13]:
```suggestion
  private static readonly releases: Record<string, Release[] | undefined> = {
```

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.ts:95]:
```suggestion
function extractRunner(runner: string): PackageDependency | null {
```

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.ts:98]:
```suggestion
    return null;
```

File: lib/modules/manager/github-actions/extract.ts
Comment (viceice) [inline lib/modules/manager/github-actions/extract.ts:103]: we should return null if depName isn't a known runner, otherwise it extracts `abc-123` ad valid dependency and later it fails with lookup warning.

you can use the datasource static values to check if it's valid, so we don't need to maintain multiple places.

**Comment by PeterNitsche** [comment]
@viceice Can you please review all changes in one go? This whole pull request processing is a tragedy and  
an imposition. Its the third time you are commenting changes which already existed from the beginning. I feel like you are not interested in getting this enhancement merged. I am not far away from closing this unmerged PR and deleting the branch.

**Comment by viceice** [comment]
> @viceice Can you please review all changes in one go? This whole pull request processing is a tragedy and  
> an imposition. Its the third time you are commenting changes which already existed from the beginning. I feel like you are not interested in getting this enhancement merged. I am not far away from closing this unmerged PR and deleting the branch. 

sorry, but I'm doing so many reviews a day in my spare time and so sometimes I don't see all issues at once.

we're now really close to finishing this PR, so please don't abandon this.

**Review by viceice — APPROVED** [review]
thank you for being patient with me 🤗

**Comment by HonkingGoose** [comment]
Thank you @PeterNitsche for your contribution! 🥳🥳🥳

**Pull request merged**
